exposed-tests/src/test/kotlin/org/jetbrains/exposed/sql/tests/shared/entities/ViaTest.kt

```diff
@@ -288,74 +286,114 @@ class ViaTests : DatabaseTestsBase() {
         }
     }
 
-    object Projects : IntIdTable("projects") {
+    // IdTable without auto-increment is used so manual ids can be inserted without excluding SQL Server
+    object Projects : IdTable<Int>("projects") {
+        override val id = integer("id").entityId()
         val name = varchar("name", 50)
+        override val primaryKey = PrimaryKey(id)
     }
-
     class Project(id: EntityID<Int>) : IntEntity(id) {
         companion object : IntEntityClass<Project>(Projects)
 
         var name by Projects.name
-        val tasks by Task via ProjectTasks
+        var tasks by TaskWithData via ProjectTasks
     }
 
-    object ProjectTasks : CompositeIdTable("project_tasks") {
+    object ProjectTasks : Table("project_tasks") {
         val project = reference("project", Projects, onDelete = ReferenceOption.CASCADE)
         val task = reference("task", Tasks, onDelete = ReferenceOption.CASCADE)
         val approved = bool("approved")
+        val sprint = integer("sprint")
 
         override val primaryKey = PrimaryKey(project, task)
-
-        init {
-            addIdColumn(project)
-            addIdColumn(task)
-        }
-    }
-
-    class ProjectTask(id: EntityID<CompositeID>) : CompositeEntity(id) {
-        companion object : CompositeEntityClass<ProjectTask>(ProjectTasks)
-
-        var approved by ProjectTasks.approved
     }
 
-    object Tasks : IntIdTable("tasks") {
+    // IdTable without auto-increment is used so manual ids can be inserted without excluding SQL Server
+    object Tasks : IdTable<Int>("tasks") {
+        override val id = integer("id").entityId()
         val title = varchar("title", 64)
+        override val primaryKey = PrimaryKey(id)
     }
-
     class Task(id: EntityID<Int>) : IntEntity(id) {
         companion object : IntEntityClass<Task>(Tasks)
 
         var title by Tasks.title
-        val approved by ProjectTasks.approved
+        var projects by ProjectWithData via ProjectTasks
+    }
+
+    class ProjectWithData(
```

Comment:
In terms of API defining of custom `InnerTableLinkEntity` looks quite verbose to me.

I expect that it's not the first variant and if all other options are not better this variant is ok.

I also tried to make something with it, and I have a feeling that it could be made in a similar way with original entity.

The thing I tried to do is to make it looks like:

```kotlin
class ProjectWithData2(
    val project: Project,
) : InnerTableLinkEntity<Int>(project) {
    val approved by ProjectTasks.approved
    val sprint by ProjectTasks.sprint

    companion object : InnerTableLinkEntityClass<Int, ProjectWithData>(Projects, ProjectTasks)
}
```

The main change here - I added `ProjectTasks` as a second paramerter of `InnerTableLinkEntityClass`. It allows to get the same data that was taken from `row` inside `createInstance()` but on the `InnerTableLinkEntityClass` level. Looks like:

```kotlin
abstract class InnerTableLinkEntityClass<WID : Comparable<WID>, out E : InnerTableLinkEntity<WID>>(
    table: IdTable<WID>,
    val intermediateTable: Table
) : EntityClass<WID, E>(table, null, null) {

fun createInstance2(entityId: EntityID<WID>, row: ResultRow?): E {
        if (row == null) error("Row is empty. How do we handle it?")
        val entity = wrapRow(row) // same as Project.wrapRow(it)
        val values = intermediateTable.columns.map { it to row[it] } // map with values it[ProjectTasks.approved], it[ProjectTasks.sprint]
// The question is what should be returned...
    }
    
}
```

Looking at the `Entity` class there is used some reflection: 

```kotlin
private val entityPrimaryCtor: KFunction<T> by lazy { 
  klass.kotlin.primaryConstructor as KFunction<T> 
}

private val entityCtor: (EntityID<ID>) -> T = entityCtor ?: { entityID -> entityPrimaryCtor.call(entityID) }
```

Probably something similar could be done with links:

To the class `InnerTableLinkEntity` I added
```kotlin
internal val extraValues: MutableMap<Column<*>, Any?> = mutableMapOf()
```

So `InnerTableLinkEntityClass` looks like:

```
abstract class InnerTableLinkEntityClass<WID : Comparable<WID>, out E : InnerTableLinkEntity<WID>>(
    table: IdTable<WID>,
    val intermediateTable: Table
) : EntityClass<WID, E>(table, null, null) {

  private val entityPrimaryCtor: KFunction<E> by lazy { klass.kotlin.primaryConstructor as KFunction<E> }
  
  private val entityCtor: (Entity<WID>) -> E =  { entity -> entityPrimaryCtor.call(entity) }
  
  fun createInstance2(entityId: EntityID<WID>, row: ResultRow?): E {
      if (row == null) error("Row is empty. How do we handle it?")
      val entity = wrapRow(row)
      val values = intermediateTable.columns.map { it to row[it] }
      val resultEntity = entityCtor(entity)
      values.forEach {
          resultEntity.extraValues[it.first] = it.second
      }
      return resultEntity
  }
}
```

It compiles but I haven't tested it, so it's just more or less an idea of how API could look like. No idea which problems I can face up with such a change...

If you already tried that way, feel free to ignore it of course)

Reply:
I can try adding a second parameter and give this shot. One of the reasons I kept moving away from a more traditional (albeit less verbose) entity with delegates was because of the questions I mentioned in the PR description. For example, let's say the above works and the new API means it looks like what you suggested above:
```kt
class ProjectWithData2(
    val project: Project,
) : InnerTableLinkEntity<Int>(project) {
   // what if user instead accidentally/purposefully made these mutable?
    var approved by ProjectTasks.approved
    var sprint by ProjectTasks.sprint

    companion object : InnerTableLinkEntityClass<Int, ProjectWithData>(Projects, ProjectTasks)
}
```

1. Using delegates means that a new link entity would be created, not by a constructor, but by `new()`? I thought maybe this would be confusing for users, because they might assume that it schedules an insert statement like a regular entity. It also means we would need to go into `new()` and adjust logic for this entity specifically.
2. What if, like above, user made the additional fields mutable? And then they assigned a new value after setting a new entity to the link field. This would be like trying to do an update, but we couldn't schedule an update like for a regular entity because it has to go through `InnerTableLink.setValue`. So what would happen? We'd have to adjust logic in `Column.setValue` to trigger an update of the `via` field?

Tbh we need this class to be some sort of entity so it fits the contract for our caches, find methods, and wrapRow etc. But it can't be too much like a traditional entity because it doesn't fit the contract for insertion/update/deletion etc. It should only exist (and be get/set) within the context of the other entity's `via` field.